serverless/internal/storage/fs/fs.go

```diff
@@ -102,21 +116,123 @@ func (fs *Storage) LogState() api.LogState {
 // UpdateState updates the stored log state.
 func (fs *Storage) UpdateState(newState api.LogState) error {
 	fs.state = newState
+	fs.nextSeq = newState.Size
 	lsRaw, err := json.Marshal(newState)
 	if err != nil {
 		return fmt.Errorf("failed to marshal LogState: %w", err)
 	}
 	oPath := filepath.Join(fs.rootDir, layout.StatePath)
 	tmp := fmt.Sprintf("%s.tmp", oPath)
-	if err := ioutil.WriteFile(tmp, lsRaw, filePerm); err != nil {
+	if err := createExcl(tmp, lsRaw); err != nil {
 		return fmt.Errorf("failed to create temporary state file: %w", err)
 	}
 	return os.Rename(tmp, oPath)
 }
 
 // Sequence assigns the given leaf entry to the next available sequence number.
+// This method will attempt to silently squash duplicate leaves, but it cannot
+// be guaranteed that no duplicate entries will exist.
 func (fs *Storage) Sequence(leafhash []byte, leaf []byte) error {
-	return errors.New("unimplemented")
+	// 1. Check for dupe leafhash
+	// 2. Write temp file
+	// 3. Hard link temp -> seq file
+	// 4. Optimistically symlink leafhash -> seq file
+
+	// Ensure the leafhash diretory structure is pressent
+	leafDir, leafFile := layout.LeafPath(fs.rootDir, leafhash)
+	if err := os.MkdirAll(leafDir, dirPerm); err != nil {
+		return fmt.Errorf("failed to make leaf directory structure: %w", err)
+	}
+	// Check for dupe leaf already present.
+	leafFQ := filepath.Join(leafDir, leafFile)
+	if _, err := os.Stat(leafFQ); os.IsExist(err) {
+		glog.Infof("Skipping dupe leafhash 0x%x", leafhash)
+		return nil
+	}
+
+	tmp := filepath.Join(fs.rootDir, fmt.Sprintf(leavesPendingPathFmt, leafhash))
+	if err := createExcl(tmp, leaf); err != nil {
+		return fmt.Errorf("unable to write temporary file: %w", err)
+	}
+	defer func() {
+		os.Remove(tmp)
+	}()
+
+	// Now try to sequence it, we may have to scan over some newly sequenced entries
+	// if Sequence has been called since the last time an Integrate/UpdateState
+	// was called.
+	for {
+		seq := fs.nextSeq
+
+		// Ensure the sequencing directory structure is present:
+		seqDir, seqFile := layout.SeqPath(fs.rootDir, seq)
+		if err := os.MkdirAll(seqDir, dirPerm); err != nil {
+			return fmt.Errorf("failed to make seq directory structure: %w", err)
+		}
+
+		// Hardlink the sequence file to the temporary file
+		seqPath := filepath.Join(seqDir, seqFile)
+		if err := os.Link(tmp, seqPath); errors.Is(err, os.ErrExist) {
+			// That sequence number is in use, try the next one
+			fs.nextSeq++
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to link seq file: %w", err)
+		}
+
+		// Finally, try to link leafhash -> seq.
+		err := os.Symlink(seqPath, leafFQ)
+		if err != nil && !errors.Is(err, os.ErrExist) {
+			return fmt.Errorf("failed to link leafdata file: %w", err)
+		}
+
+		// All done!
+		break
+	}
+
+	return nil
+}
+
+// createExcl, creates the named file before writing the data in d to it.
```

Comment:
`createExcl` took me a few turns to parse. I expanded its name to mean "this methods creates an excl.*" and started expanding this to "exclusion" etc first.

Reply:
-> `createExclusive`